src/fork.c

```diff
@@ -0,0 +1,433 @@
+		goto unlock_out;
+	}
+
+	atfork_pool.pkcss[i] = NULL;
+	atfork_pool.pkcs_num--;
+
+	_gen_free((void **)&atfork_pool.pkcss, &atfork_pool.pkcs_num,
+		  &atfork_pool.pkcs_size);
+	rc = OSSL_RV_OK;
+unlock_out:
+	if (pthread_mutex_unlock(&atfork_pool.mutex)) {
+		ps_dbg_error(dbg, "pkcs: %p, unlock atfork pool failed", pkcs);
+		return OSSL_RV_ERR;
+	}
+	/* ----- unlocked ----- */
+	ps_dbg_debug(dbg, "pkcs: %p, unregistered in atfork pool", pkcs);
+	return rc;
+}
+
+#define AFP_OH_POOL	16
+int atforkpool_register_objecthandle(CK_OBJECT_HANDLE_PTR poh, struct dbg *dbg)
+{
+	int rc = OSSL_RV_ERR;
+	bool found = false;
+	unsigned int i;
+
+	if (!poh)
+		return OSSL_RV_OK;
+	if (!dbg)
+		return OSSL_RV_ERR;
+
+	if (pthread_mutex_lock(&atfork_pool.mutex)) {
+		ps_dbg_error(dbg, "poh: %p, lock atfork pool failed", poh);
+		return OSSL_RV_ERR;
+	}
+
+	/* ----- locked ----- */
+	if (_gen_alloc((void **)&atfork_pool.ohs,
+		       &atfork_pool.oh_num, &atfork_pool.oh_size,
+		       sizeof(CK_OBJECT_HANDLE_PTR), AFP_OH_POOL) != OSSL_RV_OK) {
+		ps_dbg_error(dbg, "poh: %p, poh pool allocation failed", poh);
+		goto unlock_out;
+	}
+
+	for (i = 0; i < atfork_pool.oh_size; i++) {
+		if (atfork_pool.ohs[i] == NULL) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found) {
+		ps_dbg_error(dbg, "poh: %p, unable to register", poh);
+		goto unlock_out;
+	}
+
+	atfork_pool.ohs[i] = poh;
+	atfork_pool.oh_num++;
+
+	if (_pthread_atfork_once() != OSSL_RV_OK) {
+		ps_dbg_warn(dbg, "unable to register fork handler");
+		goto unlock_out;
+	}
+
+	rc = OSSL_RV_OK;
+unlock_out:
+	if (pthread_mutex_unlock(&atfork_pool.mutex)) {
+		ps_dbg_error(dbg, "poh: %p, unlock atfork pool failed", poh);
+		return OSSL_RV_ERR;
+	}
+	/* ----- unlocked ----- */
+	ps_dbg_debug(dbg, "poh: %p, registered in atfork pool", poh);
+	return rc;
+}
+
+int atforkpool_unregister_objecthandle(CK_OBJECT_HANDLE_PTR poh, struct dbg *dbg)
+{
+	int rc = OSSL_RV_ERR;
+	bool found = false;
+	unsigned int i;
+
+	if (!poh)
+		return OSSL_RV_OK;
+	if (!dbg)
+		return OSSL_RV_ERR;
+
+	if (pthread_mutex_lock(&atfork_pool.mutex)) {
+		ps_dbg_error(dbg, "poh: %p, lock atfork pool failed", poh);
+		return OSSL_RV_ERR;
+	}
+
+	/* ----- locked ----- */
+	for (i = 0; i < atfork_pool.oh_size; i++) {
+		if (atfork_pool.ohs[i] == poh) {
+			found = true;
+			break;
+		}
+	}
+
+	if (!found) {
+		ps_dbg_error(dbg, "poh: %p, unable to unregister", poh);
+		goto unlock_out;
+	}
+
+	atfork_pool.ohs[i] = NULL;
+	atfork_pool.oh_num--;
+
+	_gen_free((void **)&atfork_pool.ohs, &atfork_pool.oh_num,
+		  &atfork_pool.oh_size);
+	rc = OSSL_RV_OK;
+unlock_out:
+	if (pthread_mutex_unlock(&atfork_pool.mutex)) {
+		ps_dbg_error(dbg, "poh: %p, unlock atfork pool failed", poh);
+		return OSSL_RV_ERR;
+	}
+	/* ----- unlocked ----- */
+	ps_dbg_debug(dbg, "poh: %p, unregistered in atfork pool", poh);
+	return rc;
+}
+
+#define AFP_SH_POOL	16
+int atforkpool_register_sessionhandle(CK_SESSION_HANDLE_PTR psh, struct dbg *dbg)
```

Comment:
The implementation of the 3 register functions (and the 3 unregister functions as well) look very similar. They only differ in which type of thing they register in the _gen_alloc call and the loop. I wonder if  the code could not be generalized somehow, or at least the common parts of these functions could e moved into a single helper function ?

Reply:
ok, I'll check this.

Reply:
Hm, seems to be a non-trivial task. If this is OK for you, I would like to postpone the generalization of register/unregister.

Reply:
sure, leave it as is.